add a new method and a function

Freakwill opened the pull request:
powermethod is a well known method to compute the eigenvalues.

define a function to_reciprocal_matrix(A). With the help of the function, we can use 0 in a preference matrix. it will be corrected to a positive number, in order to make A reciprocal. see yettelevision.json. It makes the writing of json convenient.

Freakwill committed:
Add files via upload

add a new method

Freakwill committed:
Add files via upload

define a function to_reciprocal_matrix(A), translating A to a reciprocal matrix

Freakwill committed:
Add files via upload

yet television.

mishrabhinav commented on examples/yettelevision.json:
```diff
@@ -0,0 +1,46 @@
+      [5, 3, 0, 1]
+    ]
+  }
+}
```
Every file should have a newline at the end.

Freakwill replied:
The Power Method is a classical method to calculate the single eigenvalue with maximal abstract value, i.e. |lambda_1| > |lambda2| >= |lambda_i| in numerical analysis. You can find it in any textbook about numerical analysis. The speed of convergence is dominated by |lambda2|/|lambda1|. Moreover, it is simpler then LM method.
Perron theorem states that lambda_1>|lambda_i| for any positive matrix.

Thank you for your advice.

mishrabhinav replied:
Yes, I am very well aware of the power method to calculate the eigenvectors but my question is about the application of Power method to calculate the priorities from a preference matrix? I am looking for some kind of academic evidence here.

Freakwill replied:
`Perron theorem` is the evidence. The initial vector is recommended to be selected randomly.

mishrabhinav commented on pyahp/utils.py:
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
+                            A[i, j] = A[i, k] * A[k, j]
+                            break
+                    else:
+                        raise ValueError('There are so many zeros! I am not able to convert A[%d,%d] to a vaild value!'%(i,j))
```
The error messages should not be in first person. Please create a new error in `errors.py` (similar to the ones in the file) and raise that instead of a `ValueError`.

Freakwill replied:
ok

Freakwill replied:
I think it will be better that you define the error. Meanwhile I will think about it.

mishrabhinav commented on pyahp/methods/power.py:
```diff
@@ -0,0 +1,51 @@
+    m, n = A.shape
+    u0 = initvec or np.ones(n)
+    k = 0
+    while True:
```
Can we use the iteration count here to control the loop instead of `while True`?

Freakwill replied:
ok

mishrabhinav commented on pyahp/methods/power.py:
```diff
@@ -0,0 +1,51 @@
+        mu = v1[ind]
+        u0 = v1 / mu
+        k += 1
+        if k >= iteration or LA.norm(v0 - u0)/np.max((1, LA.norm(u0))) < tol:
```
In relation to the previous comment, this could be a `for k in range(iteration)` instead of `while True`.

mishrabhinav commented on pyahp/utils.py:
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
+
+    A = A.astype(np.float64)
+    m, n = A.shape
+    for i in range(m):
```
I am not going to lie but this loop looks terrible, so many levels of indentation! Can you try and reduce that?

Freakwill replied:
ok

mishrabhinav commented on pyahp/utils.py:
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
+    return A
+
+
+def is_reciprocal(A):
```
The `Method` abstract class has a function to check for reciprocal matrices already. Also, it doesn't make sense to add functions to the `utils` if they are not being used somewhere in the code.

Freakwill replied:
yes.

mishrabhinav commented on examples/yettelevision.json:
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "Television Model",
```
This model could use a different and more descriptive name. Same goes for the filename.

Freakwill replied:
ok

mishrabhinav commented on pyahp/methods/power.py:
```diff
@@ -0,0 +1,51 @@
+from pyahp.methods import Method
+
+
+def eig(A, initvec=None, tol=0.0001, iteration=10):
```
The implementation of the method is quite naive in my opinion as it seems to assume that it is always going to converge. I would like to see some test cases here with a good number of base, regular and edge cases.

mishrabhinav reviewed (changes requested):
I like the idea of having zeroes instead of decimals in the preference matrices and it's something I would like to ship in the next release. For the power method, do you have any supporting text/paper that it is a reliable AHP priority calculation method?

Freakwill committed:
Add files via upload

modify `to_reciprocal_matrix` in utils.py

Freakwill committed:
Add files via upload

Freakwill committed:
Delete utils.py

Freakwill committed:
Add files via upload

give an example in doc of eig function

Freakwill committed:
Add files via upload

Freakwill committed:
Add files via upload

Freakwill committed:
Update power.py

Freakwill committed:
Update utils.py

correct a typo

Freakwill committed:
Add files via upload

add a new error

Freakwill committed:
Add files via upload

Freakwill committed:
Add files via upload

Freakwill commented:
In the last commit, I considered your review as possible as I can. I don't kown modify it any more without your new review.

Do we need to define an interface for users? I am considering a fuzzy-version of AHP now. it seams uneasy.

I am attempting to use git commands at first.

mishrabhinav commented:
@Freakwill GitHub has great features and while it does allow you to upload files using the UI with a standard commit message, "Add files via upload", I would like you to change it to provide a more descriptive commit message. There are great guides online about writing commit messages and if you are interested, you could also look into using git from the command line([here](https://www.atlassian.com/git/tutorials)). Let me know if you need some help with that.

Freakwill committed:
Add files via upload

add a variable `availableMethods`. One only need modify the variable for extension, instead of affecting other files.

Freakwill committed:
Add files via upload

Freakwill committed:
Add files via upload
